dags/financial_data_dag.py

```diff
@@ -0,0 +1,121 @@
+import datetime
+from airflow import DAG
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.operators.python_operator import PythonOperator
+
+import requests
+from bs4 import BeautifulSoup
+import yfinance as yf
+import pandas as pd
+import math
+import json
+
+def get_sp500_ticker_symbols():
+    
+    url = "https://en.wikipedia.org/wiki/List_of_S%26P_500_companies"
+    r = requests.get(url,timeout=2.5)
+    r_html = r.text
+    soup = BeautifulSoup(r_html, 'html.parser')
+    
+    components_table= soup.find_all(id = "constituents")
+    data_rows = components_table[0].find("tbody").find_all("tr")[1:]
+
+    tickers = []
+    
+    for row in range(len(data_rows)):
+        stock = list(filter(None,data_rows[row].text.split("\n")))
+        symbol = stock[0]
+        
+        if (symbol.find('.') != -1):
+            symbol = symbol.replace('.', '-')
+            
+        tickers.append(symbol)
+        
+    tickers.sort()
+    
+    return tickers
+
+def download_YFinance_data_function():
+
+    tickers = get_sp500_ticker_symbols()
+    data = yf.download(tickers[:5], start = '2021-11-30')['Adj Close']
+    return data.to_json()
+
+def format_insert_function(**kwargs):
+    
+    ti = kwargs['ti']
+    
+    # pulling data (as string)
+    string_data = ti.xcom_pull(task_ids = 'download_data_task')
+    
+    # transforming to json
+    json_data = json.loads(string_data)
+    
+    # transforming to dataframe for easier manipulation
+    df = pd.DataFrame.from_dict(json_data, orient = 'index')
+
+    insert_stmt = "INSERT INTO sp500 (closing_date, ticker, adjusted_close) VALUES "
+
+    for col in range(len(df.columns)):
+        values_array = []
+        closing_date = df.columns[col]
+        
+        for row in range(len(df.index)):
+            ticker = df.index[row]
+            adj_close = df.iloc[row, col]
+            
+            if (not(adj_close is None or math.isnan(adj_close))):
+                values_array.append("({},\'{}\',{})".format(closing_date, ticker, adj_close))
+
+        insert_stmt += ", ".join(values_array) + ","
+        
+    insert_stmt = insert_stmt[:len(insert_stmt)-1] + ';'
+    print(insert_stmt)
+    
+    return  insert_stmt
```

Comment:
I don't fully understand what happens with the data inside the loop, but from what I see, the `INSERT` statement is generated, stored via XCom at the end of the task, loaded from XCom by the next task, and executed as-is. This can be an indicator of a missing separation of concerns (i.e., every task is responsible for one specific aspect). One aspect being data cleansing/preparation, the other being data ingestion (actual SQL).

In that regard, I would suggest keeping `format_insert_function` free from any SQL-specific logic and using it only for data cleansing/preparation. Then, in the next task, generate the SQL statement and execute it. This way, each function is responsible for one single concern.